02-peer-protocol.md

```diff
@@ -430,6 +446,515 @@ would create a Denial of Service risk; therefore, forgetting it is recommended
+#### Rationale
+
+Accepter sends their `funding_satoshi` value here instead of allowing the opener to derive
+it from their `funding_compose` response so that the opener can notionally decide whether
+to complete the opening without exposing their output set.
+
+`channel_reserve_satoshi` has been omitted. The channel reserve is fixed at 1% of
+the total channel balance (sum of `funding_satoshis` from `open_channel2` and `accept_channel2`) 
+or the `dust_limit_satoshis`, whichever is greater.
+
+
+### Funding Composition
+There are three messages which nodes use to communicate their input and output set
+for a funding transaction: `funding_add_input`, `funding_add_output`, and `funding_add_complete`.
+
+To add inputs to a transaction, a node may send one or more `funding_add_input` messages.
+To add outputs to a transaction, a node may send one or more `funding_add_output` messages.
+
+Once a node has relayed all of their inputs and output contributions, it sends
+`funding_add_complete` to signal the completion of their contribution transmission.
+
+#### The `funding_add_input` Message
+
+This message contains information for inputs to a funding transaction.
+
+1. type: 58 (`funding_add_input`)
+2. data:
+    * [`32*byte`:`temporary_channel_id`]
+    * [`u16`:`num_inputs`]
+    * [`num_inputs*input_info`:`input_info`]
+
+1. subtype: `input_info`
+2. data:
+    * [`u64`:`sats`]
+    * [`sha256`:`prevtx_txid`]
+    * [`u32`:`prevtx_vout`]
+    * [`u16`:`prevtx_scriptpubkey_len`]
+    * [`prevtx_scriptpubkey_len*byte`:`prevtx_scriptpubkey`]
+    * [`u16`:`max_witness_len`]
+    * [`u16`:`scriptlen`]
+    * [`scriptlen*byte`:`script`]
+
+##### Requirements
+
+The `accepter` node:
+  - MAY omit this message
+
+The sending node:
+  - MUST NOT send `funding_add_input` if it has already transmitted `funding_add_complete`
+  - MUST send this after `accept_channel2` has been exchanged, but before `funding_add_complete`
+  - MUST add all sent inputs to the funding transaction
+  - MUST NOT re-transmit inputs it has already received from the peer
+  - MUST ensure each `input_info` refers to a non-malleable (segwit) UTXO.
+  - if is the `opener`:
+    - MUST send at least one `funding_add_input` message
+    - MUST NOT send a total count of more than 64 inputs, across all `funding_add_input` messages.
+  - if is the `accepter`:
+    - MAY omit this message
+    - MUST NOT send a total count of more than 16 inputs, across all `funding_add_input` messages.
+
+The receiving node:
+  - MUST add all received inputs to the funding transaction
+  - MUST fail the channel if:
+    - it receives a duplicate input to one it sent previously
+    - if it receives this message after `funding_add_complete` is received.
+    - it receives an input that is malleable (P2SH/P2PKH)
+  - if is the `opener`:
+    - MUST fail the channel if:
+      - the total count of `input_info`s is greater than 6
+
+
+##### Rationale
+Each node must have a complete set of the transaction inputs
```

Comment:
Do we assume all nSequence to be set to SEQUENCE_FINAL ? We should at least have one of them signalling RBF.

Reply:
Good point, this is underspecified.